### ISSUE TITLE
Question: Can multiple models be saved/exported from one TPOT optimization?

### ISSUE THREAD

**Issue by pan-alex** [EpistasisLab/tpot#703]
Hi, I'm new to using TPOT, but it looks very promising. My question is if it is possible to save the parameters and/or export code for multiple models from one TPOT optimization, rather than just the most predictive model.

For example, I would like to see the top 5 (or whatever) models to assess if there are any common features, or maybe to choose a model that is easier to explain compared to one with marginal improvements in accuracy.

Thanks!

**Comment by weixuanfu** [comment]
For now, tpot will save all pipelines into `evaluated_individuals_` attributes in [TPOT API](https://epistasislab.github.io/tpot/api/). Similar issue #516 

evaluated_individuals_: Python dictionary
Dictionary containing all pipelines that were evaluated during the pipeline optimization process, where the key is the string representation of the pipeline and the value is a tuple containing (# of steps in pipeline, accuracy metric for the pipeline). 

This attribute is primarily for internal use, but may be useful for looking at the other pipelines that TPOT evaluated.

**Comment by rhiever** [comment]
You can also access the `pareto_front_fitted_pipelines_` attribute of TPOT, which provides a list of all pipelines along the Pareto front. This Pareto front shows the trade-off between pipeline predictive performance (e.g., accuracy) and pipeline complexity (i.e., the number of steps in the pipeline).

See the [TPOT API](http://epistasislab.github.io/tpot/api/) for more info on the `pareto_front_fitted_pipelines_` attribute.

**Comment by pan-alex** [comment]
Thanks @weixuanfu and @rhiever! This answered my question.



For anyone with similar needs I've included some code using the minimal working example on the MNIST data from the tpot documentation.

```
import pandas as pd
from tpot import TPOTClassifier
from sklearn.datasets import load_digits
from sklearn.model_selection import train_test_split

digits = load_digits()
X_train, X_test, y_train, y_test = train_test_split(digits.data, digits.target,
                                                    train_size=0.75, test_size=0.25)

tpot = TPOTClassifier(generations=5, population_size=50, verbosity=2)
tpot.fit(X_train, y_train)
print(tpot.score(X_test, y_test))
```

Data frame of all models and sort by best CV error:

```
my_dict = list(tpot.evaluated_individuals_.items())

model_scores = pd.DataFrame()
for model in my_dict:
    model_name = model[0]
    model_info = model[1]
    cv_score = model[1].get('internal_cv_score')  # Pull out cv_score as a column (i.e., sortable)
    model_scores = model_scores.append({'model': model_name,
                                        'cv_score': cv_score,
                                        'model_info': model_info,},
                                       ignore_index=True)

model_scores = model_scores.sort_values('cv_score', ascending=False)
```

E.g., top 5 rows of output

  | cv_score | model | model_info
-- | -- | -- | --
276 | 0.986708 | KNeighborsClassifier(DecisionTreeClassifier(input_matrix,   DecisionTreeClassifier__criterion=gini, DecisionTreeClassifier__max_depth=9,   DecisionTreeClassifier__min_samples_leaf=3,   DecisionTreeClassifier__min_samples_split=14), KNeighborsClassifier__n_neighbors=5,   KNeighborsClassifier__p=2, KNeighborsClassifier__weights=uniform) | {'generation': 'INVALID', 'mutation_count': 3, 'crossover_count': 0,   'predecessor': ('KNeighborsClassifier(input_matrix,   KNeighborsClassifier__n_neighbors=5, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=uniform)',), 'operator_count': 2, 'internal_cv_score':   0.9867084418027815}
251 | 0.986686 | KNeighborsClassifier(DecisionTreeClassifier(input_matrix,   DecisionTreeClassifier__criterion=entropy,   DecisionTreeClassifier__max_depth=3,   DecisionTreeClassifier__min_samples_leaf=18,   DecisionTreeClassifier__min_samples_split=16), KNeighborsClassifier__n_neighbors=5,   KNeighborsClassifier__p=2, KNeighborsClassifier__weights=uniform) | {'generation': 'INVALID', 'mutation_count': 1, 'crossover_count': 2,   'predecessor': ('KNeighborsClassifier(DecisionTreeClassifier(input_matrix,   DecisionTreeClassifier__criterion=entropy,   DecisionTreeClassifier__max_depth=3, DecisionTreeClassifier__min_samples_leaf=18,   DecisionTreeClassifier__min_samples_split=16),   KNeighborsClassifier__n_neighbors=19, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=uniform)', 'KNeighborsClassifier(input_matrix,   KNeighborsClassifier__n_neighbors=5, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=distance)'), 'operator_count': 2,   'internal_cv_score': 0.9866863255542502}
244 | 0.986686 | KNeighborsClassifier(GradientBoostingClassifier(input_matrix,   GradientBoostingClassifier__learning_rate=1.0,   GradientBoostingClassifier__max_depth=1,   GradientBoostingClassifier__max_features=0.15000000000000002,   GradientBoostingClassifier__min_samples_leaf=20,   GradientBoostingClassifier__min_samples_split=11,   GradientBoostingClassifier__n_estimators=100,   GradientBoostingClassifier__subsample=0.7500000000000001),   KNeighborsClassifier__n_neighbors=5, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=uniform) | {'generation': 'INVALID', 'mutation_count': 3, 'crossover_count': 0,   'predecessor': ('KNeighborsClassifier(input_matrix,   KNeighborsClassifier__n_neighbors=5, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=uniform)',), 'operator_count': 2, 'internal_cv_score':   0.9866863255542502}
202 | 0.986686 | KNeighborsClassifier(input_matrix, KNeighborsClassifier__n_neighbors=5,   KNeighborsClassifier__p=2, KNeighborsClassifier__weights=uniform) | {'generation': 'INVALID', 'mutation_count': 2, 'crossover_count': 0,   'predecessor': ('KNeighborsClassifier(RFE(input_matrix,   RFE__ExtraTreesClassifier__criterion=gini,   RFE__ExtraTreesClassifier__max_features=0.1,   RFE__ExtraTreesClassifier__n_estimators=100, RFE__step=0.6500000000000001),   KNeighborsClassifier__n_neighbors=5, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=uniform)',), 'operator_count': 1,   'internal_cv_score': 0.9866863255542502}
89 | 0.986678 | KNeighborsClassifier(input_matrix, KNeighborsClassifier__n_neighbors=5,   KNeighborsClassifier__p=2, KNeighborsClassifier__weights=distance) | {'generation': 'INVALID', 'mutation_count': 1, 'crossover_count': 0,   'predecessor': ('KNeighborsClassifier(RFE(input_matrix,   RFE__ExtraTreesClassifier__criterion=gini,   RFE__ExtraTreesClassifier__max_features=0.1,   RFE__ExtraTreesClassifier__n_estimators=100, RFE__step=0.6500000000000001),   KNeighborsClassifier__n_neighbors=5, KNeighborsClassifier__p=2,   KNeighborsClassifier__weights=distance)',), 'operator_count': 1,   'internal_cv_score': 0.9866781855461101}

**Comment by Karim-53** [comment]
code obtained in https://github.com/EpistasisLab/tpot/issues/703#issuecomment-395632097 cannot be used directly.
Please use the code below to obtain a ready to use version of the code (like in tpot.export(...) ) for the entire pareto

```
from tpot.export_utils import (
    export_pipeline,
    expr_to_tree,
    generate_pipeline_code,
    set_param_recursive,
)

# tpot = TPOTRegressor(generations=25, ...)
# tpot.fit()


for pipeline, pipeline_scores in zip(tpot._pareto_front.items, reversed(tpot._pareto_front.keys)):
    idx = tpot._pareto_front.items.index(pipeline)
    pareto_front_pipeline_score = pipeline_scores.wvalues[1]
    sklearn_pipeline_str = generate_pipeline_code(
        expr_to_tree(pipeline, tpot._pset), tpot.operators
    )
    to_write = export_pipeline(
        pipeline,
        tpot.operators,
        tpot._pset,
        tpot._imputed,
        pareto_front_pipeline_score,
        tpot.random_state,
    )
    print(to_write, '\n----------------------------------------------------------------------------\n\n')
```

example of output:

```
import numpy as np
import pandas as pd
from sklearn.ensemble import GradientBoostingRegressor
from sklearn.model_selection import train_test_split

# NOTE: Make sure that the outcome column is labeled 'target' in the data file
tpot_data = pd.read_csv('PATH/TO/DATA/FILE', sep='COLUMN_SEPARATOR', dtype=np.float64)
features = tpot_data.drop('target', axis=1)
training_features, testing_features, training_target, testing_target = \
            train_test_split(features, tpot_data['target'], random_state=53)

# Average CV score on the training set was: -0.21667382164465937
exported_pipeline = GradientBoostingRegressor(alpha=0.85, learning_rate=0.1, loss="ls", max_depth=6, max_features=0.9000000000000001, min_samples_leaf=1, min_samples_split=13, n_estimators=100, subsample=0.8)
# Fix random state in exported estimator
if hasattr(exported_pipeline, 'random_state'):
    setattr(exported_pipeline, 'random_state', 53)

exported_pipeline.fit(training_features, training_target)
results = exported_pipeline.predict(testing_features)
 
---------------------------------------------------------------------------- -0.21667382164465937 


import numpy as np
import pandas as pd
from sklearn.linear_model import LassoLarsCV
from sklearn.model_selection import train_test_split
from sklearn.pipeline import make_pipeline, make_union
from tpot.builtins import StackingEstimator
from xgboost import XGBRegressor
from tpot.export_utils import set_param_recursive

# NOTE: Make sure that the outcome column is labeled 'target' in the data file
tpot_data = pd.read_csv('PATH/TO/DATA/FILE', sep='COLUMN_SEPARATOR', dtype=np.float64)
features = tpot_data.drop('target', axis=1)
training_features, testing_features, training_target, testing_target = \
            train_test_split(features, tpot_data['target'], random_state=53)

# Average CV score on the training set was: -0.19100860580801965
exported_pipeline = make_pipeline(
    StackingEstimator(estimator=LassoLarsCV(normalize=False)),
    XGBRegressor(learning_rate=0.1, max_depth=3, min_child_weight=6, n_estimators=100, n_jobs=1, objective="reg:squarederror", subsample=1.0, verbosity=0)
)
# Fix random state for all the steps in exported pipeline
set_param_recursive(exported_pipeline.steps, 'random_state', 53)

exported_pipeline.fit(training_features, training_target)
results = exported_pipeline.predict(testing_features)
 
---------------------------------------------------------------------------- -0.19100860580801965 

and 2 more items below...
```

hope this help
